book/18-concurrent-programming.md

```diff
@@ -0,0 +1,1062 @@
+the dominant approach in languages like Java or C#. In this model, each task
+that may require simultaneous waiting is given an operating system thread of
+its own so it can block without stopping the entire program.
+<idx>threads/preemptive vs. single-threaded programs</idx>
+
+ Another approach is to have a single-threaded program, where that single
+thread runs an *event loop* whose job is to react to external events like
+timeouts or mouse clicks by invoking a callback function that has been
+registered for that purpose. This approach shows up in languages like
+JavaScript that have single-threaded runtimes, as well as in many GUI
+toolkits. <idx>event loops</idx><idx>system threads</idx>
+
+ Each of these mechanisms has its own trade-offs. System threads require
+significant memory and other resources per thread. Also, the operating system
+can arbitrarily interleave the execution of system threads, requiring the
+programmer to carefully protect shared resources with locks and condition
+variables, which is exceedingly error-prone.
+
+ Single-threaded event-driven systems, on the other hand, execute a single
+task at a time and do not require the same kind of complex synchronization
+that preemptive threads do. However, the inverted control structure of an
+event-driven program often means that your own control flow has to be
+threaded awkwardly through the system's event loop, leading to a maze of
+event callbacks.
+
+ This chapter covers the Async library, which offers a hybrid model that aims
+to provide the best of both worlds, avoiding the performance compromises and
+synchronization woes of preemptive threads without the confusing inversion of
+control that usually comes with event-driven systems. <idx>Async
+library/benefits of</idx>
+
+## Async Basics {#async-basics data-type="sect1"}
+
+ Recall how I/O is typically done in Core. Here's a simple example:
+
+<link rel="import" href="code/async/main.mlt" part="1" />
+
+ From the type of `In_channel.read_all`, you can see that it must be a
+blocking operation. In particular, the fact that it returns a concrete string
+means it can't return until the read has completed. The blocking nature of
+the call means that no progress can be made on anything else until the call
+is complete. <idx>blocking</idx>
+
+ In Async, well-behaved functions never block. Instead, they return a value
+of type `Deferred.t` that acts as a placeholder that will eventually be
+filled in with the result. As an example, consider the signature of the Async
+equivalent of `In_channel.read_all`. <idx>Deferred.t</idx>
+
+<link rel="import" href="code/async/main.mlt" part="3" />
+
+ We first load the Async package in the toplevel using `#require`, and then
+open the module. Async, Like Core, is designed to be an extension to your
+basic programming environment, and is intended to be opened.
+
+ A deferred is essentially a handle to a value that may be computed in the
```

Comment:
Is the leading space in front of each graf important? Or is it just an artifact of the conversion?